### PR TITLE
MINOR; Update scalafmt to latest version

### PR DESCRIPTION
**Pull request by mdedetrich**
Updates scalafmt to the latest version. This not only means migrating
the checkstyle/.scalafmt.conf file but updating spotless as well. Due
to improvements in the new version of scalafmt, it now also
alphabetically sorts imports.

The diff to the actual .scala source files is due to the alphabetical sorting of imports. I verified that this PR passes spotless/scalacheck by running ` ./gradlew :spotlessScalaCheck` locally.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: build.gradle
Comment (mdedetrich) [inline build.gradle]: This version needed to be updated otherwise spotless would bring resolve the wrong Scala version for scalafmt.

File: checkstyle/.scalafmt.conf
Comment (mdedetrich) [inline checkstyle/.scalafmt.conf:17]: This keeps the current docstrings intact (i.e. forces scalafmt to ignore docstrings). Turning it on messes with manual html formatting that is used in docstrings

File: checkstyle/.scalafmt.conf
Comment (mdedetrich) [inline checkstyle/.scalafmt.conf:18]: This forces scalafmt to ignore docstrings when formatting code. The new version of scalafmt works on docstrings however it doesn't support manual html formatting which is done in kafka scala streams.

**Comment by mdedetrich** [comment]
@divijvaidya This is one of the first PR's in adding the linting/formatting using scalafmt/scalafix for the Scala sources within Kafka that you talked about here https://github.com/apache/kafka/pull/12471#issuecomment-1202378703. @cadonna Do you want to have a look as well since its part of kafka scala streams?

File: checkstyle/.scalafmt.conf
Comment (divijvaidya) [inline checkstyle/.scalafmt.conf:16]: We still support scala 2.12 in this project.

By specifying the dialect as 2.13, would we fail where syntax for 2.12 is used? If yes, are there any such cases in the project? (The build should tell us that after the existing spotless check passes).

**Review by divijvaidya — COMMENTED** [review]
Thanks for making this change. 

1. Note that build is failing related to changes in this PR. We would probably need to fix the error but I think I might have already fixed it in https://github.com/apache/kafka/pull/12230 If that lands, then the check should pass here.
```
[2022-08-02T22:58:41.580Z] * What went wrong:

[2022-08-02T22:58:41.580Z] Execution failed for task ':spotlessScala'.

[2022-08-02T22:58:41.580Z] > java.lang.reflect.InvocationTargetException
```

2. I have verified that the [style guide we follow](https://docs.scala-lang.org/style/scaladoc.html#general-style) for this project is not prescriptive about formatting of the docs. Hence, your change should be ok.

File: checkstyle/.scalafmt.conf
Comment (mdedetrich) [reply checkstyle/.scalafmt.conf:16]: The Scala 2.13 dialect is backwards compatible with 2.12. I have applied the dialect when updating from scalafmt 2 to scalafmt 3 with various popular OSS Scala projects and there isn't any issues (in any case the Scala code that is used im Kafka is quite conservative)

> I have verified that the style guide we follow for this project is not prescriptive about formatting of the docs. Hence, your change should be ok.

Do you think it makes sense to format the docs or should I leave the PR as is?

> Note that build is failing related to changes in this PR. We would probably need to fix the error but I think I might have already fixed it in MINOR: Catch InvocationTargetException explicitly and propagate underlying cause #12230 If that lands, then the check should pass here.

In that case it's sensible to wait for your PR to land and then I can rebase.

Comment (divijvaidya) [reply checkstyle/.scalafmt.conf:16]: > Do you think it makes sense to format the docs or should I leave the PR as is?

I think we should leave the PR as is. Converting the docs in the entire project will be cumbersome without any benefit. 

> In that case it's sensible to wait for your PR to land and then I can rebase.

Thanks, in that case, would you kindly review that PR please if you get some time? It's been pending for a while now waiting to get some attention.

Comment (mdedetrich) [reply checkstyle/.scalafmt.conf:16]: > Thanks, in that case, would you kindly review that PR please if you get some time? It's been pending for a while now waiting to get some attention.

I just landed a review of the PR.

**Comment by ijuma** [comment]
The scala 2.12 build failed https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-12475/

**Comment by mdedetrich** [comment]
@ijuma Thanks for the pickup.

I have investigated the root cause of the problem, basically when you specify the scala version with `-PscalaVersion=2.12` it also happens to override the Scala version that is used by scalafmt and the latest version of scalafmt (3.5.8) only runs with Scala 2.13.x. This is normally a non issue because scalafmt is designed to be used like a cli tool, i.e. its meant to be run independently of your project.

I created an issue upstream at https://github.com/diffplug/spotless/issues/1273

**Comment by mdedetrich** [comment]
@ijuma So I have added a commit which adds a workaround for the Scala runtime conflict issues. The CI is failing but it appears it's not using the updated `Jenkinsfile` in the commit so I suspect that after a merge the Apache CI should use the updated Jenkinsfile?

@mimaison For visibility as well.

**Comment by mdedetrich** [comment]
Actually lets wait on merging this PR, it may be possible to fix the underlying problem in spotless (see https://github.com/diffplug/spotless/issues/1273#issuecomment-1222221116)

**Comment by mdedetrich** [comment]
@ijuma So I have some good news, I managed to completely the solve the underlying problem upstream (see https://github.com/diffplug/spotless/issues/1273#issuecomment-1224978741). The PR has been updated/rebased with the newest spotless and as you can see `spotlessScalaCheck` even works when Scala 2.12 is set, i.e.

```sh
./gradlew -PscalaVersion=2.12 :spotlessScalaCheck
```

Also tagging @mimaison for visibility

**Comment by mdedetrich** [comment]
Tagging @C0urante for visibility

File: checkstyle/.scalafmt.conf
Comment (C0urante) [inline checkstyle/.scalafmt.conf:15]: Why are we specifying this version here? Isn't it already declared in `dependencies.gradle`?

**Review by C0urante — COMMENTED** [review]
Thanks Matthew! One small question about the changes.

I should note that I wasn't a fan of the import shuffling initially, but it does seem like we wanted import sorting when we originally introduced Scala formatting in https://github.com/apache/kafka/pull/4965 and included the `SortImports` rule. So although someone else more familiar with this part of the code base may feel differently, at least from my perspective these changes are fine.

I've kicked off another Jenkins run to see if we can get at least one green build with Scala 2.12.

File: checkstyle/.scalafmt.conf
Comment (mdedetrich) [reply checkstyle/.scalafmt.conf:15]: So I checked and its actually a requirement of the newer versions of Scalafmt, if you don't have the version there it complains with

```
Caused by: java.util.NoSuchElementException: version [expected 3.5.9 or 3.5.9]: missing
        at metaconfig.Configured.get(Configured.scala:15)
        at com.diffplug.spotless.glue.scalafmt.ScalafmtFormatterFunc.<init>(ScalafmtFormatterFunc.java:46)
        at java.base/jdk.internal.reflect.DirectConstructorHandleAccessor.newInstance(DirectConstructorHandleAccessor.java:67)
        ... 130 more
```

Not sure what the reasoning behind having to include the include the scalafmt version in the configuration file but aside from some hacky workaround it needs to be done this away.

Comment (C0urante) [reply checkstyle/.scalafmt.conf:15]: Hmmm... it's a little strange that we now have to specify the version in two places (`gradle/dependencies.gradle` and `checkstyle/.scalafmt.conf`), and it might lead to some headaches when we want to upgrade to a newer version in the future.

If there's no sane way to define the Scalafmt version in a single place (preferably `gradle/dependencies.gradle`), could we leave a comment in the dependencies file noting that the version will also have to be bumped in the Scalafmt configuration file?

FWIW, I checked out the history for Scalafmt to see if we could get some additional context there, but the [PR](https://github.com/scalameta/scalafmt/pull/2843) that added the requirement for the `version` property didn't shed a lot of light.

Comment (mdedetrich) [reply checkstyle/.scalafmt.conf:15]: So I found the release notes where this change was added, i.e. https://github.com/scalameta/scalafmt/releases/tag/v3.1.0 but unfortunately like the PR its not very informative as to why the version now has to be explicit.

In any case I have rebased the PR by adding a comment notifying that whenever `scalafmt` is bumped in `dependencies.gradle` that it also needs to be changed in `checkstyle/.scalafmt.conf` as requested.

**Comment by mdedetrich** [comment]
> I've kicked off another Jenkins run to see if we can get at least one green build with Scala 2.12.

So Jenkins did in fact pass this PR as green build. I have just pushed to this PR so another build has been triggered but https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka-pr/detail/PR-12475/8/pipeline is the build that passed.

File: checkstyle/.scalafmt.conf
Comment (C0urante) [reply checkstyle/.scalafmt.conf:15]: Thanks, I think that should be fine.

**Review by C0urante — APPROVED** [review]
LGTM! I'd like it if someone more familiar with Streams could also take a look before merging, though.

**Comment by C0urante** [comment]
@vvcephei @cadonna would either of you have time to double-check this quick Scalafmt bump?

**Pull request merged**
